stl/inc/span

```diff
@@ -0,0 +1,665 @@
+    using element_type           = _Ty;
+    using value_type             = remove_cv_t<_Ty>;
+    using index_type             = size_t;
+    using difference_type        = ptrdiff_t;
+    using pointer                = _Ty*;
+    using const_pointer          = const _Ty*;
+    using reference              = _Ty&;
+    using const_reference        = const _Ty&;
+    using iterator               = _Span_iterator<_Ty>;
+    using const_iterator         = _Span_iterator<const _Ty>;
+    using reverse_iterator       = _STD reverse_iterator<iterator>;
+    using const_reverse_iterator = _STD reverse_iterator<const_iterator>;
+
+    static constexpr index_type extent = _Extent;
+
+    // [span.cons] Constructors, copy, and assignment
+    constexpr span() noexcept = default;
+
+    constexpr span(pointer _Ptr, index_type _Count) noexcept // strengthened
+        : _Mybase(_Count), _Mydata(_Ptr) {}
+
+    constexpr span(pointer _First, pointer _Last) noexcept // strengthened
+        : _Mybase(static_cast<index_type>(_Last - _First)), _Mydata(_First) {
+#if _CONTAINER_DEBUG_LEVEL > 0
+        _STL_VERIFY(_First <= _Last, "span construction from pointers only valid for first <= last");
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+    }
+
+    template <size_t _Size, enable_if_t<_Extent == dynamic_extent || _Extent == _Size, int> = 0>
+    constexpr span(element_type (&_Arr)[_Size]) noexcept : _Mybase(_Size), _Mydata(_Arr) {}
+
+    template <size_t _Size, enable_if_t<_Extent == dynamic_extent || _Extent == _Size, int> = 0>
+    constexpr span(array<value_type, _Size>& _Arr) noexcept : _Mybase(_Size), _Mydata(_Arr.data()) {}
+
+    template <size_t _Size, enable_if_t<_Extent == dynamic_extent || _Extent == _Size, int> = 0>
+    constexpr span(const array<value_type, _Size>& _Arr) noexcept : _Mybase(_Size), _Mydata(_Arr.data()) {}
+
+    template <class _Cont,
+        enable_if_t<_Is_span_compatible_container_v<_Cont, element_type> && _Extent == dynamic_extent, int> = 0>
+    constexpr span(_Cont& _Container)
+        : _Mybase(static_cast<index_type>(_STD size(_Container))), _Mydata(_STD data(_Container)) {}
+
+    template <class _Cont,
+        enable_if_t<_Is_span_compatible_container_v<const _Cont, element_type> && _Extent == dynamic_extent, int> = 0>
+    constexpr span(const _Cont& _Container)
+        : _Mybase(static_cast<index_type>(_STD size(_Container))), _Mydata(_STD data(_Container)) {}
+
+    // clang-format off
+    template <class _OtherTy, size_t _OtherExtent,
+        enable_if_t<is_convertible_v<_OtherTy (*)[], element_type (*)[]> 
+				&& (_Extent == dynamic_extent || _Extent == _OtherExtent), int> = 0>
+    constexpr span(const span<_OtherTy, _OtherExtent>& _Other) noexcept
+        : _Mybase(_Other.size()), _Mydata(_Other.data()) {}
+    // clang-format on
+
+    // [span.sub] Subviews
+    template <size_t _Count>
+    _NODISCARD constexpr span<element_type, _Count> first() const noexcept // strengthened
+    {

```

Comment:
Thanks for updating the `strengthened` comments. One more stylistic nitpick: we attempt to keep `strengthened` as close as possible to `noexcept`, but we don't like detached braces, so we're willing to use C comments to avoid detachment. See https://github.com/microsoft/STL/blob/957fe99f41748b80c3108447e5dd5ce2be962f4e/stl/inc/array#L164 for an example. The `strengthened` comments in the constructors above are fine; they can be C++ comments because the next line begins with a mem-initializer-list. But for these ordinary functions, `noexcept /* strengthened */ {` is our current convention. (Perhaps this is overly confusing and should be globally changed, but that's what @BillyONeal had settled on last time this was overhauled.)

Reply:
Done